src/python/pants/backend/javascript/package_json.py

```diff
@@ -440,6 +444,83 @@ class NodeBuildScriptTarget(Target):
     )
 
 
+@dataclass(frozen=True)
+class PackageJsonImports:
+    """https://nodejs.org/api/packages.html#subpath-imports."""
+
+    imports: FrozenDict[re.Pattern[str], tuple[str, ...]]
+    root_dir: str
+
+    def replacements(self, import_string: str) -> tuple[str, ...]:
+        def replace_matching_pattern(
+            pattern: re.Pattern[str], subpath: str, string: str
+        ) -> str | None:
+            match = pattern.match(string)
+            if match:
+                replacement = subpath
+                for group in match.groups():
+                    replacement = replacement.replace("*", group, 1)
+                if "*" in replacement:
+                    _logger.warning(
+                        softwrap(
+                            f"""
+                            package.json#imports pattern '{pattern.pattern}' matched '{string}',
+                            but the resulting subpath '{subpath}' string replacements '*'
+                            did not match.
+
+                            Inference will not behave correctly for import '{string}'.
+                            """
+                        )
+                    )
+                    return None
+                return "".join((replacement, string[match.endpos :]))
+            return None
+
+        return tuple(
+            filter(
+                None,
+                (
+                    replace_matching_pattern(pattern, subpath, import_string)
+                    for pattern, subpaths in self.imports.items()
+                    for subpath in subpaths
+                ),
+            )
+        )
+
+    @classmethod
+    def from_package_json(cls, pkg_json: PackageJson) -> PackageJsonImports:
+        return cls(
+            imports=cls._import_from_package_json(pkg_json),
+            root_dir=pkg_json.root_dir,
+        )
+
+    @staticmethod
+    def _to_import_pattern(string: str) -> re.Pattern[str]:
+        return re.compile(r"^" + re.escape(string).replace(r"\*", "(.*)"))
```

Comment:
Are these supposed to match files recursively, or only within a directory? If the latter, you might want to use `([^/]*)`?